### PR TITLE
ISPN-6580 Hotrod performance regressions after ISPN-5342 ISPN-6545

### PR DESCRIPTION
**Pull request by gustavonalle**
https://issues.jboss.org/browse/ISPN-6580

Some numbers from the ```RemoteClientPutGetTest```, on Linux (server run with ```-Dinfinispan.server.channel.epoll=true```) 

||8.2.1.Final | 9.0.0-SNAPSHOT
----|-------|-----------------------
|Put| 13s | 15s 
|Get| 71s | 66s
|Alloc| 130 MB/s | 47 MB/s

### CODE REVIEW

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/Reproducer.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
+package org.infinispan.client.hotrod;
+
+import org.infinispan.client.hotrod.configuration.ConfigurationBuilder;
+
+import java.util.Random;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+
+public class Reproducer {
```
Comment (anistor) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/Reproducer.java]: Looks like an improper test; not located under src/test/ and name is not like *Test.java.

File: server/core/src/main/scala/org/infinispan/server/core/transport/ExtendedByteBufJava.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+import io.netty.buffer.ByteBuf;
+import io.netty.util.CharsetUtil;
+import org.infinispan.commons.io.SignedNumeric;
+
+/**
+ * // TODO: Document this
+ *
+ * @author wburns
+ * @since 4.0
```
Comment (anistor) [inline server/core/src/main/scala/org/infinispan/server/core/transport/ExtendedByteBufJava.java]: 4.0? (same in almost all other files)

File: server/core/src/main/scala/org/infinispan/server/core/transport/ExtendedByteBufJava.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
+package org.infinispan.server.core.transport;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.util.CharsetUtil;
+import org.infinispan.commons.io.SignedNumeric;
+
+/**
+ * // TODO: Document this
```
Comment (anistor) [inline server/core/src/main/scala/org/infinispan/server/core/transport/ExtendedByteBufJava.java]: yes, please! (or just remove the todo :))

File: server/core/src/main/scala/org/infinispan/server/core/transport/ExtendedByteBufJava.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+ *
+ * @author wburns
+ * @since 4.0
+ */
+public class ExtendedByteBufJava {
```
Comment (anistor) [inline server/core/src/main/scala/org/infinispan/server/core/transport/ExtendedByteBufJava.java]: This deserves a private constructor.

File: server/hotrod/src/main/scala/org/infinispan/server/hotrod/transport/HotRodChannelInitializer.scala [hunk 5]
```diff
@@ -16,19 +19,24 @@ class HotRodChannelInitializer(val server: HotRodServer, transport: => NettyTran
...
 
       // Logging handlers
-      ch.pipeline.addBefore("decoder", "logging", new HotRodAccessLoggingHandler)
-      ch.pipeline.addAfter("encoder", "logging-context", LoggingContextHandler.getInstance);
+      if (LogFactory.getLog(classOf[HotRodAccessLoggingHandler], classOf[JavaLog]).isTraceEnabled) {
+         ch.pipeline.addBefore("decoder", "logging", new HotRodAccessLoggingHandler)
+         ch.pipeline.addAfter("encoder", "logging-context", LoggingContextHandler.getInstance)
+      }
    }
 }
```
Comment (anistor) [inline server/hotrod/src/main/scala/org/infinispan/server/hotrod/transport/HotRodChannelInitializer.scala]: Feels incomplete :)

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/Reproducer.java (hunk 1 above)
Comment (gustavonalle) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/Reproducer.java]: Ok, will move it under "profiling" group on test folder

File: server/hotrod/src/main/scala/org/infinispan/server/hotrod/transport/HotRodChannelInitializer.scala (hunk 5 above)
Comment (gustavonalle) [inline server/hotrod/src/main/scala/org/infinispan/server/hotrod/transport/HotRodChannelInitializer.scala]: Why?

**Comment by gustavonalle** [comment]
There is one test failure on ```SiteDownFailoverTest```, investigating...

**Comment by gustavonalle** [comment]
Test failure fixed.

File: server/hotrod/src/main/scala/org/infinispan/server/hotrod/transport/HotRodChannelInitializer.scala (hunk 5 above)
Comment (anistor) [inline server/hotrod/src/main/scala/org/infinispan/server/hotrod/transport/HotRodChannelInitializer.scala]: Was referring to the missing EOL :)

**Comment by gustavonalle** [comment]
@anistor Rebased and added the EOL :smile:

**Comment by gustavonalle** [comment]
Updated again. Pushed one tiny fix to ```RemoteClientPutGetTest```

**Comment by gustavonalle** [comment]
Rebased. Also a slight change of plan: epoll is disabled by default for now, and can be enabled by a System property.

**Comment by tristantarrant** [comment]
Pushed. Great job @gustavonalle and @wburns

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/ContextHandler.java [hunk 6]
```diff
@@ -33,14 +39,34 @@
...
    }
 
    @Override
    protected void channelRead0(ChannelHandlerContext ctx, CacheDecodeContext msg) throws Exception {
+      executor.execute(() -> {
+         try {
+            Subject subject = msg.getSubject();
+            if (subject == null)
+               realRead(ctx, msg);
+            else Security.doAs(subject, (PrivilegedExceptionAction<Void>) () -> {
```
Comment (wburns) [inline server/hotrod/src/main/java/org/infinispan/server/hotrod/ContextHandler.java]: I was actually thinking what might be cleaner was instead to wrap the provided executor in the ChannelInitializer when security is enabled.  This way security is completely isolated from the other code.  We also then wouldn't need any if checks etc.

File: server/hotrod/src/main/scala/org/infinispan/server/hotrod/transport/HotRodChannelInitializer.scala [hunk 7]
```diff
@@ -16,19 +19,24 @@ class HotRodChannelInitializer(val server: HotRodServer, transport: => NettyTran
...
-      ch.pipeline.addLast(executionGroup, "exception", new HotRodExceptionHandler)
+      ch.pipeline.addLast("local-handler", new LocalContextHandler(transport))
+
+      ch.pipeline.addLast("handler", new ContextHandler(server, transport, executor))
+      ch.pipeline.addLast("exception", new HotRodExceptionHandler)
 
       // Logging handlers
-      ch.pipeline.addBefore("decoder", "logging", new HotRodAccessLoggingHandler)
-      ch.pipeline.addAfter("encoder", "logging-context", LoggingContextHandler.getInstance);
+      if (LogFactory.getLog(classOf[HotRodAccessLoggingHandler], classOf[JavaLog]).isTraceEnabled) {
```
Comment (wburns) [inline server/hotrod/src/main/scala/org/infinispan/server/hotrod/transport/HotRodChannelInitializer.scala]: Unfortunately doing this means that if tracing is enabled dynamically after the server is started only new connections will have the access log.  Is this okay @tristantarrant ?  This is why I had the trace enabled checks in the handlers themselves.

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/ContextHandler.java (hunk 6 above)
Comment (wburns) [inline server/hotrod/src/main/java/org/infinispan/server/hotrod/ContextHandler.java]: @tristantarrant WDYT about this?

Comment (tristantarrant) [inline server/hotrod/src/main/java/org/infinispan/server/hotrod/ContextHandler.java]: Security checks would be per-cache while the transport is per-container. So you'd need to check anyway

**Pull request closed without merging**